API: Make datetime64 timezone naive

shoyer opened the pull request:
Fixes #3290

With apologies to @mwiebe, this rips out most of the time zone parsing from the datetime64 type.

I think we mostly sorted out the API design in [discussions last year](http://thread.gmane.org/gmane.comp.python.numeric.general/57184), but nonetheless I'll be posting this to the mailing list shortly to get feedback.

Old behavior:

```
# string parsing and printing defaults to your local timezone :(
>>> np.datetime64('2000-01-01T00')
numpy.datetime64('2000-01-01T00:00-0800','h')
```

New behavior:

```
# datetime64 is parsed and printed as timezone naive
>>> np.datetime64('2000-01-01T00')
numpy.datetime64('2000-01-01T00','h')

# you can still supply a timezone, but you get a deprecation warning
>>> np.datetime64('2000-01-01T00Z')
DeprecationWarning: parsing timezone aware datetimes is deprecated; this
will raise an error in the future
numpy.datetime64('2000-01-01T00','h')
```

Still be resolved: How should we handle the function `np.datetime_as_string`? As far as I can tell, it was never advertised in the public API and mostly exists for testing, but it still exists in the main numpy namespace :(. If we can remove it, then we can delete and simplify a lot more code related to timezone parsing and display. If not, we'll need to do a bit of work so we can distinguish between the string representations of timezone naive and UTC.

cc @PythonCHB @jreback

jreback commented on numpy/core/arrayprint.py:
```diff
@@ -708,9 +708,9 @@ def __call__(self, x):
 class DatetimeFormat(object):
-    def __init__(self, x, unit=None,
-                timezone=None, casting='same_kind'):
+    def __init__(self, x, unit=None, timezone='naive', casting='same_kind'):
```
not sure what the standard for numpy is, but usually ``timezone=None`` is by definition naive.

PythonCHB replied:
Agreed -- ``None`` should mean naive time.

mwiebe replied:
The lack of a naive time zone was probably the biggest problem in the datetime NEP as it was when I started doing any datetime work. Unfortunately I didn't realize it should be added, and the discussion at the time didn't help with that.

shoyer replied:
OK, `timezone=None` now implies `timezone='naive'`

jreback commented:
lgtm from pandas point of view, fixes the 'display issue', which confounds users.

PythonCHB commented:
@shoyer : Thank you SO MUCH for doing this -- it's been a long time!

PythonCHB commented:
On Sun, Oct 11, 2015 at 11:51 PM, Stephan Hoyer <notifications@github.com>
wrote:

> # you can still supply a timezone, but you get a deprecation warning
>
> >>> np.datetime64('2000-01-01T00Z')
> DeprecationWarning: parsing timezone aware datetimes is deprecated; this
> will raise an error in the future
> numpy.datetime64('2000-01-01T00','h')
>
> This is a bit odd -- so it's kinda-sorta UTC? i.e if you supply an offset,
you get UTC? but then it's completely naive from there on out?

*sigh* -- I suppose we need to keep that for backwards compat. -- but I
wonder if the lack of any timezone on the other end will break it anyway?

Does anyone have use-cases, where they pass in offsets and need UTC? But
not any other TZ handling?

I'm a bit tempted to only allow "Z" or a 00:00 offset (though still
deprecate that), as that really makes no difference.

-CHB







> Still be resolved: How should we handle the function np.datetime_to_string?
> As far as I can tell, it was never advertised in the public API and mostly
> exists for testing, but it still exists in the main numpy namespace :(. If
> we can remove it, then we can delete and simplify a lot more code related
> to timezone parsing and display. If not, we'll need to do a bit of work so
> we can distinguish between timezone naive and UTC (which should be denoted
> by ending with Z).
>
> cc @PythonCHB <https://github.com/PythonCHB> @jreback
> <https://github.com/jreback>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/numpy/numpy/pull/6453
> Commit Summary
>
>    - WIP: Make datetime64 timezone naive
>
> File Changes
>
>    - *M* numpy/core/arrayprint.py
>    <https://github.com/numpy/numpy/pull/6453/files#diff-0> (14)
>    - *M* numpy/core/src/multiarray/datetime_strings.c
>    <https://github.com/numpy/numpy/pull/6453/files#diff-1> (107)
>    - *M* numpy/core/src/multiarray/scalartypes.c.src
>    <https://github.com/numpy/numpy/pull/6453/files#diff-2> (24)
>    - *M* numpy/core/tests/test_datetime.py
>    <https://github.com/numpy/numpy/pull/6453/files#diff-3> (128)
>
> Patch Links:
>
>    - https://github.com/numpy/numpy/pull/6453.patch
>    - https://github.com/numpy/numpy/pull/6453.diff
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6453>.
>



-- 
Christopher Barker, PhD

Python Language Consulting
  - Teaching
  - Scientific Software Development
  - Desktop GUI and Web Development
  - wxPython, numpy, scipy, Cython

shoyer commented:
@PythonCHB Yes, this would preserve the legacy behavior of converting datetimes with explicitly provided timezones into UTC. I think this is the least painful way to transition from the current behavior, which is something like "UTC only, but prints in local time".

PythonCHB commented:
On Tuesday, October 13, 2015, Stephan Hoyer <notifications@github.com>
wrote:

> @PythonCHB <https://github.com/PythonCHB> Yes, this would preserve the
> legacy behavior of converting datetimes with explicitly provided timezones
> into UTC.
>
And probably necessary-- as I posted on the Numpy list -- with the current
code, you HAVE to specify UTC in an iOS string to not get any offset
applied.

-CHB




>
>
> I think this is the least painful way to transition from the current
> behavior, which is something like "UTC only, but prints in local time".
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6453#issuecomment-147857017>.
>


-- 
Christopher Barker, PhD

Python Language Consulting
  - Teaching
  - Scientific Software Development
  - Desktop GUI and Web Development
  - wxPython, numpy, scipy, Cython

shoyer commented:
OK, the latest patch preserves `np.datetime_as_string` (perhaps we should deprecate it, but I don't think there's cause to eliminate it immediately), though the default timezone treatment has shifted to naive.

charris commented:
@shoyer Could you add a description of this to the 1.11 release notes, squash the commits, fix up the commit messages. @mwiebe Mark, could you a quick look at this?

mwiebe commented on numpy/core/src/multiarray/datetime_strings.c:
```diff
@@ -811,17 +656,21 @@ parse_iso_8601_datetime(char *str, Py_ssize_t len,
+        /* by default, we no longer interpret times as local */
         if (out_local != NULL) {
-            *out_local = 1;
+            *out_local = 0;
```
Might be better to rename the parameter from `out_local` to `out_naive`, keep it set to 1 here, and update the documentation comment for the function, etc.

shoyer replied:
Actually, maybe we just entirely drop this parameter instead? AFAICT, it is not actually used anywhere.

mwiebe commented on numpy/core/src/multiarray/datetime_strings.c:
```diff
@@ -1075,7 +924,7 @@ lossless_unit_from_datetimestruct(npy_datetimestruct *dts)
 NPY_NO_EXPORT int
 make_iso_8601_datetime(npy_datetimestruct *dts, char *outstr, int outlen,
-                    int local, NPY_DATETIMEUNIT base, int tzoffset,
+                    int local, int utc, NPY_DATETIMEUNIT base, int tzoffset,
```
Probably would be good to add information about the `utc` parameter to the doc comment.

shoyer replied:
Added a note.

I'm not really happy with defining a new variable `utc` here -- I'd rather use a string or some sort of enum type as the function parameter -- but this was the easiest thing to do (given my poor knowledge of C).

mwiebe commented on numpy/core/tests/test_datetime.py:
```diff
@@ -1794,21 +1827,22 @@ def test_datetime_is_busday(self):
+            a = np.datetime64('2038-01-19T04:14:08+0100')
+            assert_equal(a.view(np.int64), 2**31)
 
         # Test parsing a date after Y2038 in the local timezone
```
Comment would be more clear as "in the naive timezone." Could do a global search/audit of comments with the phrase "local timezone" and tweak them to be consistent with this pull request's change.

mwiebe commented:
The change looks reasonable to me. The way I would describe this is turning the datetime from "strongly typed" as UTC only (with bugs around the edges) into "weakly typed" via a naive time zone. Probably the main lesson from experience is that the datetime64 design should have included a naive timezone.

shoyer commented on numpy/core/tests/test_datetime.py:
```diff
@@ -20,6 +21,14 @@
 
 
+@contextlib.contextmanager
+def assert_deprecated():
```
do we have an existing context manager that does this check?

charris replied:
No, but this isn't correct, `w` is empty if the test fails and you will get an `IndexError`. We should fix up `assert_warns` so that it works. The problem (I think) is that `DeprecationWarning` and a few others are turned into errors for development testing. We usually do something like
```
    with warnings.catch_warnings():
        warnings.simplefilter('error', DeprecationWarning)
        assert_raises(DeprecationWarning, ...)
```
but that is clumsy. Let me take a look at `assert_warns`.

charris replied:
Oh, and the stack trace is somewhat extreme. We should fix that also.

charris replied:
OK, `assert_warns` should work in this case.

charris replied:
There is also a `test_deprecations.py` file that helps us track deprecations.

shoyer commented:
The last commit includes fixes per @mwiebe's review, but also two other changes:
- documentation for the release notes and updates to the datetime64 docs.
- casting from dates to times is now allowed (e.g., `np.datetime64('2000-01-01', 'h')`, because the distinction is no longer relevant with time-zone naive datetimes.

charris commented:
@mwiebe Thanks for taking a look.

charris commented:
Going for the trifecta here. Just needs a bit more work...

shoyer commented:
Yep, I'll have to take another look at this later tonight -- I have some
actual work to do now :)

On Thu, Jan 14, 2016 at 3:36 PM, Charles Harris <notifications@github.com>
wrote:

> Going for the trifecta here. Just needs a bit more work...
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/numpy/numpy/pull/6453#issuecomment-171817473>.
>

shoyer committed:
TST: Make assert_warns an optional contextmanager

shoyer commented:
@charris oops -- looks like my rebase ate your comments. But I think I have addressed them.

Note: I extended `assert_warns` to work like a context manager. It's just very awkward to use it with a callable in code that actually does something.

shoyer committed:
API: Make datetime64 timezone naive

Fixes GH3290

With apologies to mwiebe, this rips out most of the time zone parsing from
the datetime64 type.

I think we mostly sorted out the API design in discussions last year, but
I'll be posting this to the mailing list shortly to get feedback.

Old behavior:

	# string parsing and printing defaults to your local timezone :(
    >>> np.datetime64('2000-01-01T00')
    numpy.datetime64('2000-01-01T00:00-0800','h')

New behavior:

    # datetime64 is parsed and printed as timezone naive
    >>> np.datetime64('2000-01-01T00')
    numpy.datetime64('2000-01-01T00','h')

    # you can still supply a timezone, but you get a deprecation warning
    >>> np.datetime64('2000-01-01T00Z')
    DeprecationWarning: parsing timezone aware datetimes is deprecated; this
    will raise an error in the future
    numpy.datetime64('2000-01-01T00','h')

charris commented on numpy/core/tests/test_datetime.py:
```diff
@@ -1,5 +1,6 @@
 from __future__ import division, absolute_import, print_function
 
+import contextlib
```
Don't need this any more.

shoyer replied:
fixed

charris commented:
I like the context manager idea, that new option should be mentioned in the release notes and the documentation strings should mention that the context manager functionality is new in 1.11. There is also an unneeded import of `contextlib`. Apart from that, LGTM. We should probably add the context manager option to `assert_raises` at some point.

charris commented:
Oh, in the deprecation checks, please add a comment that the deprecation is new in 1.11, doing that helps us keep track of when things were deprecated so that they can be removed or changed to errors at some future time.

shoyer commented:
assert_raises actually already works as a context manager -- at least if you are using Python 2.7 or newer. I'm not quite sure why that is but we wrap nose for the functionality currently. In any case that should take care of itself when we drop 2.6 support -- which will be soon, I think?

On Fri, Jan 15, 2016 at 8:45 AM, Charles Harris <notifications@github.com>
wrote:

> I like the context manager idea, that new option should be mentioned in the release notes and the documentation strings should mention that the context manager functionality is new in 1.11. There is also an unneeded import of `contextlib`. Apart from that, LGTM. We should probably add the context manager option to `assert_raises` at some point.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/numpy/numpy/pull/6453#issuecomment-172011755

charris commented:
We'll drop 2.6 right after the 1.11.x branch.

shoyer commented:
Do we really want to announce new features in NumPy's internal test utilities? If anyone is using `assert_warns` outside of numpy they are probably doing it wrong...

shoyer commented:
@charris I already have a note in the docstring for `TestDatetime64Timezone` that it is deprecated in 1.11.

charris commented:
They aren't numpy internal ;) If you grep scipy there are lots of uses and I expect there are other projects using them as well.

shoyer commented:
> They aren't numpy internal ;)

Only because NumPy doesn't really have such a thing as private API ;). But fair enough.

shoyer committed:
DOC: document changes to assert_warns

shoyer commented:
Just pushed release notes and version changed documentation for `assert_warns`.

charris commented:
Thanks Stephan.
